src/core/providers/qgsprovidersublayertask.cpp

```diff
@@ -0,0 +1,50 @@
+/***************************************************************************
+                             qgsprovidersublayertask.cpp
+                             ----------------------
+    begin                : June 2021
+    copyright            : (C) 2021 by Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsprovidersublayertask.h"
+#include "qgsfeedback.h"
+#include "qgsproviderregistry.h"
+#include "qgsprovidersublayerdetails.h"
+
+QgsProviderSublayerTask::QgsProviderSublayerTask( const QString &uri )
+  : QgsTask( tr( "Retrieving layers" ), QgsTask::CanCancel | QgsTask::CancelWithoutPrompt )
+  , mUri( uri )
+{
+}
+
+QList<QgsProviderSublayerDetails> QgsProviderSublayerTask::results() const
+{
+  return mResults;
+}
+
+QgsProviderSublayerTask::~QgsProviderSublayerTask() = default;
+
+bool QgsProviderSublayerTask::run()
+{
+  mFeedback = std::make_unique< QgsFeedback >();
+
+  mResults = QgsProviderRegistry::instance()->querySublayers( mUri, Qgis::SublayerQueryFlag::ResolveGeometryType | Qgis::SublayerQueryFlag::CountFeatures, mFeedback.get() );
```

Comment:
isn't there a risk for an extermely small race if someone would call results() while run() is setting mResults ?

Reply:
Fixed